### DISCUSSION TITLE
string 타입 데이터는 어떻게 메소드를 사용할 수 있나요?

### DISCUSSION THREAD

**Discussion by javascriptbiggosoo** [wanted-9th-3team/frontend-study#38] [프론트 기술 질문 공간]
알다시피 _array_ 타입과 _string_ 타입은 `includes` 같은 자바스크립트 내장 메소드를 사용할 수 있습니다.
_prototype_ 을 배운다면 자바스크립트에서 _array_ 타입은 사실 _object_ 타입으로서, 결국 상위 _prototype_ 에서 메소드를 상속받아 사용한다는 사실을 알게 되는데요.
```jsx
const strs = ['A','B','C']
console.log(strs) 
/*
(3) ['A', 'B', 'C']
0
: 
"A"
1
: 
"B"
2
: 
"C"
length
: 
3
[[Prototype]]
: 
Array(0)
$fastPatch
: 
1
at
: 
ƒ at()
concat
: 
ƒ concat()
constructor
: 
ƒ Array()
copyWithin
: 
ƒ copyWithin()
entries
: 
ƒ entries()
every
: 
ƒ every()
fill
: 
ƒ fill()
filter
: 
ƒ filter()
find
: 
ƒ find()
findIndex
: 
ƒ findIndex()
findLast
: 
ƒ findLast()
findLastIndex
: 
ƒ findLastIndex()
flat
: 
ƒ flat()
flatMap
: 
ƒ flatMap()
forEach
: 
ƒ forEach()
includes
: 
ƒ includes()
indexOf
: 
ƒ indexOf()
join
: 
ƒ join()
keys
: 
ƒ keys()
lastIndexOf
: 
ƒ lastIndexOf()
length
: 
0
map
: 
ƒ map()
pop
: 
ƒ (...t)
push
: 
ƒ (...t)
reduce
: 
ƒ reduce()
reduceRight
: 
ƒ reduceRight()
reverse
: 
ƒ (...t)
shift
: 
ƒ (...t)
slice
: 
ƒ slice()
some
: 
ƒ some()
sort
: 
ƒ (...t)
splice
: 
ƒ (...t)
toLocaleString
: 
ƒ toLocaleString()
toReversed
: 
ƒ toReversed()
toSorted
: 
ƒ toSorted()
toSpliced
: 
ƒ toSpliced()
toString
: 
ƒ toString()
unshift
: 
ƒ (...t)
values
: 
ƒ values()
with
: 
ƒ with()
Symbol(Symbol.iterator)
: 
ƒ values()
Symbol(Symbol.unscopables)
: 
{at: true, copyWithin: true, entries: true, fill: true, find: true, …}
[[Prototype]]
: 
Object
constructor
: 
ƒ Object()
hasOwnProperty
: 
ƒ hasOwnProperty()
isPrototypeOf
: 
ƒ isPrototypeOf()
propertyIsEnumerable
: 
ƒ propertyIsEnumerable()
toLocaleString
: 
ƒ toLocaleString()
toString
: 
ƒ toString()
valueOf
: 
ƒ valueOf()
__defineGetter__
: 
ƒ __defineGetter__()
__defineSetter__
: 
ƒ __defineSetter__()
__lookupGetter__
: 
ƒ __lookupGetter__()
__lookupSetter__
: 
ƒ __lookupSetter__()
__proto__
: 
(...)
get __proto__
: 
ƒ __proto__()
set __proto__
: 
ƒ __proto__()*/
```
그렇다면 _object_ 타입이 아닌 _string_ 타입의 경우 메소드가 없는데 어떻게 메소드를 사용할 수 있는 것일까요?
```jsx
const str = 'ABC';
console.log(str) // ABC
```
![image](https://user-images.githubusercontent.com/28939278/233036987-61c73699-0d1e-49ef-ac12-d12fb855571c.png)

**Comment by michoball** [comment]
생각해보지 않은 부분이네요! 그냥 되길래 당연하게 사용했는데 말이죠 ㅋㅋㅋㅋ

제가 가상으로 string 타입의 변수를 하나 만들고 __proto__로 타고 올라가니 String이라는 객체 안에 prototype으로 includes 메서드가 선언되어 있네요. 

그리고 ECMAScript 공식 문서에 보니 두 메서드의 이름은 같은데 로직에 좀 차이가 있는것 같습니다. 

어렵네요 ㅠㅠ 같이 살펴보고 의견 나눠보면 좋을 것 같습니다. 

https://tc39.es/ecma262/#sec-array.prototype.includes
https://tc39.es/ecma262/#sec-string.prototype.includes

**Comment by javascriptbiggosoo** [comment]
method가 실행되는 순간 자동으로 string primitive는 string object로 변환됩니다.
바로 그 상태에서 method를 가져다 쓰고는 value가 return되면 다시 string을 primitive로 돌려놓습니다.
이 과정을 boxing이라고 부릅니다.


```jsx
const 진영 = '진영'
진영.slice(1) // new String('진영').slice(1)
```



**레퍼런스 타입이 아닌 primitive 타입인 string, number, 불리언들도 **intermediate object wrappers**를 갖고 그곳(이 경우 string prototype)에서 모든 문자열(넘버, 불리언) 메서드를 받아온다.**

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String

**Comment by bona373737** [comment]
찾아보니 이부분을 이해하기 위해선 "wrapper 객체"에 대한 이해가 필요한것 같습니다. 
프로토타입에 대한 부분까지 연결되어 이해하려니 쉽지 않네요. 
하단에 이해에 도움이 될만한 내용 참조했습니다.  

[원시값을 객체처럼 사용하기] (https://ko.javascript.info/primitives-methods#ref-832)
자바스크립트 창안자(creator)는 다음과 같은 모순적인 상황을 해결해야만 했었습니다.

문자열이나 숫자와 같은 원시값을 다루어야 하는 작업이 많은데, 메서드를 사용하면 작업을 수월하게 할 수 있을 것 같다는 생각이 듭니다.
그런데 원시값은 가능한 한 빠르고 가벼워야 합니다.
조금 어색해 보이지만, 자바스크립트 창안자는 아래와 같은 방법을 사용해 해결책을 모색하였습니다.

원시값은 원시값 그대로 남겨둬 단일 값 형태를 유지합니다.
문자열, 숫자, 불린, 심볼의 메서드와 프로퍼티에 접근할 수 있도록 언어 차원에서 허용합니다.
이를 가능하게 하기 위해, 원시값이 메서드나 프로퍼티에 접근하려 하면 추가 기능을 제공해주는 특수한 객체, "원시 래퍼 객체(object wrapper)"를 만들어 줍니다. 이 객체는 곧 삭제됩니다.
"래퍼 객체"는 원시 타입에 따라 종류가 다양합니다. 각 래퍼 객체는 원시 자료형의 이름을 그대로 차용해, String,Number,Boolean, Symbol라고 부릅니다.
